# Pangenome: seqwish
Tags: #pangenome #pggb #seqwish #smoothxg #wfmash
ðŸ  [home](0.01_GitHub_PGGB.md)
***
> [! info] Purpose
> Seqwish uses the all-vs-all alignments to build a pangenome.

## split wfmash per chromosome

Both alignments and fasta have to be subset otherwise it creates a single node id for each chr when the fasta is not subset.

```bash
# subset wfmash
for chr in {01..19}; do awk -v chr="chr${chr}" '$1 ~ chr && $6 ~ chr' wfmash/all.on.all.wfmash_s10000p85n1.paf > wfmash/all.on.all.wfmash_s10000p85n1.chr${chr}.paf; done

# subset fasta
for chr in {01..19}; do grep "chr${chr}" ref/all.fasta | sed 's:>::g' > ref/chr${chr}.id; done

for chr in {01..19}; do seqtk subseq ref/all.fasta ref/chr${chr}.id > ref/chr${chr}.fasta; done
```

## seqwish pangenome construction

https://github.com/ekg/seqwish

k = 49 gives a good compression of the graph topology without loosing too much gene level info (core gene contains at least 50% of core sequence).

```bash
export kval=49

for chr in {01..19}; do echo "/usr/bin/time -v -o logs/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.timelog seqwish -p wfmash/all.on.all.wfmash_s10000p85n1.chr${chr}.paf -s ref/chr${chr}.fasta -g seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.gfa -t 48 -k ${kval} -P 2> logs/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.err"; done > seqwish_commands.sh

srun -N 1 -c 48 -t 96:00:00 -p bmh --mem=400G --job-name=seqwish --mail-type=ALL --mail-user=ndcochetel@ucdavis.edu seqwish_commands.sh
```

[wfmash <- Previous Step](0.02_GitHub_PGGB_wfmash.md) | [Next Step -> smoothxg](0.04_GitHub_PGGB_smoothxg.md)
