# Pangenome: seqwish
Tags: #pangenome #pggb #seqwish #smoothxg #wfmash
ðŸ  [home](README.md)
***
> [! info] Purpose
> Seqwish uses the all-vs-all alignments to build a pangenome.

## split wfmash per chromosome

Alignment and fasta files need to be filtered per chromosome to be able to run it. For the alignments, the filter is applied on the query AND the reference.

```bash
# subset wfmash
for chr in {01..19}; do awk -v chr="chr${chr}" '$1 ~ chr && $6 ~ chr' wfmash/all.on.all.wfmash_s10000p85n1.paf > wfmash/all.on.all.wfmash_s10000p85n1.chr${chr}.paf; done

# subset fasta
for chr in {01..19}; do grep "chr${chr}" ref/all.fasta | sed 's:>::g' > ref/chr${chr}.id; done

for chr in {01..19}; do seqtk subseq ref/all.fasta ref/chr${chr}.id > ref/chr${chr}.fasta; done
```

## seqwish pangenome construction

https://github.com/ekg/seqwish

k = 49 (the default) gives a good compression of the graph topology without loosing too much gene level info (core gene contains at least 50% of core sequence).

```bash
export kval=49

for chr in {01..19}; do echo "/usr/bin/time -v -o logs/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.timelog seqwish -p wfmash/all.on.all.wfmash_s10000p85n1.chr${chr}.paf -s ref/chr${chr}.fasta -g seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.gfa -t 48 -k ${kval} -P 2> logs/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.err"; done > seqwish_commands.sh

srun -N 1 -c 48 -t 96:00:00 -p bmh --mem=400G --job-name=seqwish --mail-type=ALL --mail-user=ndcochetel@ucdavis.edu seqwish_commands.sh
```

## statistics
### summary stats
```bash
for chr in {01..19}; do vg stats -z seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.gfa > seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.stats; done
```

### extract fasta per path
Needed to evaluate the size of the different pangenome classes.

```bash
# extract fasta
for chr in {01..19}; do awk 'BEGIN{FS="\t";OFS="\t"} $1=="S"' seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.gfa | awk 'BEGIN{FS="\t";OFS="\t"}  {gsub("S", ">s", $1); print $1$2"\n"$3}' > fasta/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.fasta; done

# extract len
for chr in {01..19}; do /DATA9/Resources/Scripts/getLengthFromFasta.py fasta/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.fasta > fasta/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.fasta.len; done
```

### extract paths 
Paths have to be extracted. Then, when we calculate the frequencies, we know that 2 means 2 genomes.

```bash
# extract paths with strand
for chr in {01..19}; do awk '$1=="P"' seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.gfa | cut -f 2-3 > path/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.path; done

# extract names
for chr in {01..19}; do cut -f 1 path/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.path > path/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.id; done

# from the different chr graph, extract paths and concatenate per haplotype
for chr in {01..19}; do tot_len=$(wc -l path/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.id | cut -f 1 -d ' '); for line_nb in `seq 1 ${tot_len}`; do name=$(sed -n "${line_nb}p" path/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.id); echo "sed -n "${line_nb}p" path/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.path | cut -f 2 | tr ',' '\n' | sed 's:+:\\t+:g' | sed 's:-:\\t-:g' | sed 's:^:s:g'> path/${name}.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.path"; done; done > extract_path.sh

parallel -j 24 :::: extract_path.sh

# concatenate per genome
for chr in {01..19}; do for genome in $(cat genome_loc.txt); do name=$(basename ${genome}); echo "cat path/${name}.hap[1-2].chr${chr}.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.path | cut -f 1 | sort -u > path/${name}.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.path"; done; done > path_concat_perGenome.sh

parallel -j 24 :::: path_concat_perGenome.sh
```

### class paths

```bash
# concatenate all genomes paths
for chr in {01..19}; do for genome in $(cat genome_loc.txt); do name=$(basename ${genome}); cat path/${name}.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.path; done > path/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.path; done

# get freq per node
for chr in {01..19}; do cat path/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.path | sort | uniq -c > path/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.freq; done

# class
# core
for chr in {01..19}; do awk '$1==9' path/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.freq > path/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.core.freq; done

# disp
for chr in {01..19}; do awk '$1 < 9 && $1 > 1' path/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.freq > path/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.disp.freq; done

# priv
for chr in {01..19}; do awk '$1==1' path/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.freq > path/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.priv.freq; done
```

### stats per class checkpoint
```bash
# get fasta len per class
for chr in {01..19}; do for i in $(ls path/allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.*.freq); do class=$(basename $i .freq | sed "s:allGenome.thru.all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.::g"); grep -wFf <(sed 's:^.*s:s:g' $i) fasta/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.fasta.len > fasta/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.${class}.fasta.len; done; done
```


[wfmash <- Previous Step](0.02_GitHub_PGGB_wfmash.md) | [Next Step -> smoothxg](0.04_GitHub_PGGB_smoothxg.md)
