# Pangenome: smoothxg
Tags: #pangenome #pggb #seqwish #smoothxg #wfmash
ðŸ  [home](README.md)
***
> [! info] Purpose
> The smoothxg polishing of the pangenome allows to increase the compression level of the pangenome by solving nodes with a length < k-mer size selected. It will create more nodes, but will reduce the overall size of the pangenome.

https://github.com/pangenome/smoothxg

At the time of running it, smoothxg was not intended to be run as a standalone tool, so no manual/pipeline exists. The key is to look how the pggb automatic pipeline calls smoothxg.

### round 1
```bash
for chr in {01..19}; do smoothxg -t 48 -T 48 -g ../seqwish/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k49.gfa -w 68017 -K -X 100 -I 0.85 -R 0 -j 0 -e 0 -l 4001 -P "1,4,6,2,26,1" -O 0.03 -Y 1700 -d 0 -D 0 -m all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k49.presmooth.maf -Q "Consensus_" -V -o all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k49.presmooth.gfa 2> all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k49.presmooth.log; done
```

### round 2
```bash
for chr in {01..19}; do smoothxg -t 48 -T 48 -g all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k49.presmooth.gfa -w 76619 -K -X 100 -I 0.85 -R 0 -j 0 -e 0 -l 4507 -P "1,4,6,2,26,1" -O 0.03 -Y 1700 -d 0 -D 0 -m all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k49.smooth.maf -Q "Consensus_" -V -o all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k49.smooth.gfa 2> all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k49.smooth.log; done
```

> [! summary] Summary
> The pangenomes per chromosome are ready, now the ID space can be corrected for further use.

## convert to vg

https://github.com/vgteam/vg/wiki/Index-Construction#indexes-for-vg-map-with-a-large-graph

```bash
export kval=49

for chr in {01..19}; do vg view --gfa-in smoothxg/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.smooth.gfa --vg --threads 48 > vg/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.smooth.vg 2> logs/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.smooth.gfa2vg.err; done
```

## correct IDs space

https://github.com/vgteam/vg/wiki/Index-Construction#with-haplotypes-or-with-many-paths

With 1 vg per chr, the id space is overlapping, step needed to correct for it.

```bash
cd vg

cp *vg backups/.

vg ids --join $(ls *seqwish_k49.smooth.vg)
```

## verify the joined graphs

```bash
for chr in {01..19}; do vg view --gfa --threads 48 vg/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.smooth.vg > vg/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.smooth.join.gfa 2> logs/all.on.all.wfmash_s10000p85n1.chr${chr}.seqwish_k${kval}.smooth.join.vg2gfa.err; done
```

[Previous Step -> seqwish](0.03_GitHub_PGGB_seqwish.md)
