# Pangenome: PGGB pangenome modeling
Tags: #pangenome #pggb #seqwish #smoothxg #wfmash
ðŸ  [home](5.00_Pangenome_Index.md)
***
> [! info] Purpose
> Model the gene pangenome.

## create beds

To intersect the pangenome nodes with any annotations, we need first to get a bed files describing the node composition of each genome.

[PANGENOME.smoothxg_chr_path2bed_4server.R](scripts/PANGENOME.smoothxg_chr_path2bed_4server.R)

## merge chromosome bed per genome

```bash
for genome in $(cat genome_loc.txt); do name=$(basename ${genome}); echo "cat bed/${name}.hap*.chr*thru.all.on.all.wfmash_s10000p85n1.chr*.seqwish_k49.smooth.join.bed > bed/${name}.thru.all.on.all.wfmash_s10000p85n1.seqwish_k49.smooth.join.bed"; done > merge_bed.smooth.join.sh

parallel -j 24 :::: merge_bed.smooth.join.sh
```

## intersect nodes with genes per genome

For each genome, an `awk '$3=="gene"'` was performed on the gene annotation gff3.

> [! attention]
> Do not run intersect in parallel

```bash
# intersect
for genome in $(cat genome_loc.txt); do name=$(basename ${genome}); bedtools intersect -wo -a /DATA8/Projects/North_American_Species/assembly/perGenome/haplosync/all/final_assembly/${name}.gene.gff3 -b bed/${name}.thru.all.on.all.wfmash_s10000p85n1.seqwish_k49.smooth.join.bed > intersect/${name}.thru.all.on.all.wfmash_s10000p85n1.seqwish_k49.smooth.join.intersect_gene.txt; done

# simplify
for genome in $(cat genome_loc.txt); do name=$(basename ${genome}); echo "awk 'BEGIN{FS=\"\t\";OFS=\"\t\"} {gsub(/;Name.*\$/, \"\", \$9); gsub(/;Description.*\$/, \"\", \$9); gsub(/ID=/, \"\", \$9); print \$9,\$13,\$16}' intersect/${name}.thru.all.on.all.wfmash_s10000p85n1.seqwish_k49.smooth.join.intersect_gene.txt > intersect/${name}.thru.all.on.all.wfmash_s10000p85n1.seqwish_k49.smooth.join.intersect_gene_ez.txt"; done > simplify_intersect_smooth_join.sh

parallel -j 9 :::: simplify_intersect_smooth_join.sh
```

## create all possible combinations of genomes

[PANGENOME.prep_genomes_combinations_4server.R](scripts/PANGENOME.prep_genomes_combinations_4server.R)

## extract class stats per combination

```bash
cd /DATA13/Projects/ncochtl_sva

export path_name="/DATA8/Projects/North_American_Species/assembly/pangenome"

# for modeling i = 2
# for final stats i = 9

# concatenate the path file for the combination
for comb in $(ls ${path_name}/comb/genome_nb${i}*txt); do comb_name=$(basename ${comb} .txt); cat ${comb} | while read line; do cat ${path_name}/path/${line}.thru.all.on.all.wfmash_s10000p85n1.chr*.seqwish_k49.smooth.join.path; done > path/${comb_name}.path; done

# get the freq for the combination
for path in $(ls path/genome_nb${i}*path); do comb_name=$(basename ${path} .path); sort ${path} | uniq -c > path/${comb_name}.freq; done

# reformat the freq
for freq in $(ls path/genome_nb${i}*freq); do comb_name=$(basename ${freq} .freq); echo "paste <(sed 's:s.*\$::g' ${freq} | sed 's: ::g') <(sed 's:^.*s:s:g' ${freq} | sed 's: ::g') > path/${comb_name}_freq.txt"; done > reformat_freq.sh

parallel -j 9 :::: reformat_freq.sh

# classify (each process takes up to 10 Gb of RAM)
for freq in $(ls path/genome_nb${i}*_freq.txt); do comb_name=$(basename ${freq} _freq.txt); echo "grep -wFf <(awk '\$1==1' ${freq} | cut -f 2) ${path_name}/fasta/all.on.all.wfmash_s10000p85n1.seqwish_k49.smooth.join.fasta.len > path/${comb_name}.priv.len; grep -wFf <(awk -v "i=$i" '\$1>1 && \$1<i' ${freq} | cut -f 2) ${path_name}/fasta/all.on.all.wfmash_s10000p85n1.seqwish_k49.smooth.join.fasta.len > path/${comb_name}.disp.len; grep -wFf <(awk -v "i=$i" '\$1==i' ${freq} | cut -f 2) ${path_name}/fasta/all.on.all.wfmash_s10000p85n1.seqwish_k49.smooth.join.fasta.len > path/${comb_name}.core.len"; done > class_freq.sh

parallel -j 9 :::: class_freq.sh

# stats
for freq in $(ls path/genome_nb${i}*.len); do comb_name=$(basename ${freq} .len); echo "paste <(wc -l ${freq} | cut -f 1 -d ' ') <(awk '{s+=\$2}END{print s}' ${freq}) > path/${comb_name}.stats"; done > get_stats.sh

parallel -j 9 :::: get_stats.sh

# merge
for stats in $(ls path/*stats); do name=$(basename ${stats} .stats); genome_nb=$(echo ${name} | sed 's:genome_nb::g' | sed 's:\..*$::g'); comb=$(echo ${name} | sed 's:^.*comb::g' | sed 's:\..*$::g'); type=$(echo ${name} | sed 's:^.*\.::g'); n=$(cat ${stats} | cut -f 1); len=$(cat ${stats} | cut -f 2); echo -e "${genome_nb}\t${comb}\t${type}\t${n}\t${len}"; done > path/all_stats.txt
```

## modeling

[PANGENOME.gene_intersect_smoothxg_path_modeling_4server.R](scripts/PANGENOME.gene_intersect_smoothxg_path_modeling_4server.R)

[smoothxg <- Previous Step](0.04_GitHub_PGGB_smoothxg.md) 

