# Pangenome: annotate variants
Tags: #pangenome #pggb #vg #snpeff
ðŸ  [home](README.md)
***
> [! info] Purpose
> Run snpEff on the inferred variants to evaluate their impact of the genes.

## prepare the config files

### template file
```bash
snpEff template:

#-------------------------------------------------------------------------------
# Databases & Genomes
#
# One entry per genome version.
#
# For genome version 'ZZZ' the entries look like
#       ZZZ.genome              : Real name for ZZZ (e.g. 'Human')
#       ZZZ.reference           : [Optional] Comma separated list of URL to site/s where information for building ZZZ database was extracted.
#       ZZZ.chrName.codonTable  : [Optional] Define codon table used for chromosome 'chrName' (Default: 'codon.Standard')
#
#-------------------------------------------------------------------------------

#---
# Non-standard Databases
#---

# genome_name, version version_name
genome_name.genome : genome_name genome
```

### prepare config file for each genome at chromosome level
```bash
for genome in $(ls seq/*fasta.split/*fasta.gz); do name=$(basename ${genome} .fasta.gz); version_name=$(echo ${name} | sed 's:^.*_v::' | sed 's:.hap.*$::'); sed "s:genome_name:${name}:g" snpEff_template.config > ${name}_snpEff.config; sed -i "s:version_name:${version_name}:" ${name}_snpEff.config; done
```

## prepare directories
```bash
# extract gtf and sequences.fa
for genome in $(ls seq/*fasta.split/*fasta.gz); do name=$(basename ${genome} .fasta.gz); genome_name=$(echo ${name} | sed 's:.hap.*$::'); genome_loc=$(grep ${genome_name} genome_loc.txt); echo "Preparing directories for ${name}"; mkdir -p data/${name}; echo "Converting gff3 to gtf"; gffread <(grep ${name} ${genome_loc}.gff3) -T -o data/${name}/genes.gtf; echo "Copying fasta"; gunzip -c ${genome} > data/${name}/sequences.fa; done

# extract protein.fa
for genome in $(ls seq/*fasta.split/*fasta.gz); do name=$(basename ${genome} .fasta.gz); genome_name=$(echo ${name} | sed 's:.hap.*$::'); genome_loc=$(grep ${genome_name} genome_loc.txt); grep ${name} ${genome_loc}.protein.fasta | sed 's:>::g' > data/${name}/protein.id; seqtk subseq ${genome_loc}.protein.fasta data/${name}/protein.id > data/${name}/protein.fa; done

# extract cds.fa
for genome in $(ls seq/*fasta.split/*fasta.gz); do name=$(basename ${genome} .fasta.gz); genome_name=$(echo ${name} | sed 's:.hap.*$::'); genome_loc=$(grep ${genome_name} genome_loc.txt); grep ${name} ${genome_loc}.CDS.fasta | sed 's:>::g' > data/${name}/cds.id; seqtk subseq ${genome_loc}.CDS.fasta data/${name}/cds.id > data/${name}/cds.fa; done
```

## build the database
```bash
for genome in $(ls seq/*fasta.split/*fasta.gz); do name=$(basename ${genome} .fasta.gz); genome_name=$(echo ${name} | sed 's:.hap.*$::'); genome_loc=$(grep ${genome_name} genome_loc.txt); echo "Building database for ${name}"; java -jar snpEff.jar build -gtf22 -c ${name}_snpEff.config -dataDir data -v ${name} >logs/${name}.snpEff_build.log 2> logs/${name}.snpEff_build.err; done
```

## run snpeff
```bash
for var in snp ins del mnp other; do for genome in $(ls seq/*fasta.split/*fasta.gz); do name=$(basename ${genome} .fasta.gz); genome_name=$(echo ${name} | sed 's:.hap.*$::'); genome_loc=$(grep ${genome_name} genome_loc.txt); chr_name=$(echo ${name} | sed 's:^.*chr:chr:'); echo "Annotating the ${var} of ${name}..."; java -jar snpEff.jar ann -c ${name}_snpEff.config -nodownload -dataDir data -upDownStreamLen 3000 -stats results/${name}.${var}.snpEff_summary.html ${name} ../deconstruct/parse/all.on.all.wfmash_s10000p85n1.${chr_name}.seqwish_k49.smooth.on.${name}_ae.lv0.norm.${var}.vcf > results/all.on.all.wfmash_s10000p85n1.${chr_name}.seqwish_k49.smooth.on.${name}_ae.lv0.norm.${var}.ann.vcf 2> logs/all.on.all.wfmash_s10000p85n1.${chr_name}.seqwish_k49.smooth.on.${name}_ae.lv0.norm.${var}.ann.err; done; done
```

## extract impacts and genes
```bash
# simplify the files except for others
for file in $(ls results/*ann.vcf | grep -v other); do name=$(basename ${file} .vcf); less -S ${file} | cut -d '|' -f 2-4 > results/${name}.ez.txt; done

for file in $(ls results/*ez.txt | grep -v other); do sed -i -n '/^#CHROM/,$p' ${file}; done

# others
# some of the files have no annotation, so no '|', those lines have to be discarded
for file in $(ls results/*ann.vcf | grep other); do name=$(basename ${file} .vcf); echo "sed '/#/d' ${file} | grep '|' | cut -f 8 | cut -d '|' -f 2-4 > results/${name}.ez.txt"; done > snpEff_other_commands.sh

parallel -j 24 :::: snpEff_other_commands.sh
```

## summarize the results
`PANGENOME.snpEff_SV_class.R`

[Infer variants <- Previous Step](0.07_infer_variants.md) | [home](README.md) | [Next Step -> Variant genotyping](0.09_variant_genotyping.md)
